File: cpp/LeetCode/No1137.cpp
/*
The Tribonacci sequence Tn is defined as follows: 

T0 = 0, T1 = 1, T2 = 1, and Tn+3 = Tn + Tn+1 + Tn+2 for n >= 0.

Given n, return the value of Tn.
*/
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int tribonacci(int n) {
        if(n<2) return n;
        if(n==2) return 1;
        vector<int> dp(n+1);
        dp[0] = 0;
        dp[1] = dp[2] = 1;
        for(int i=3;i<=n;i++){
            dp[i] = dp[i-3] + dp[i-2] + dp[i-1];
        }
        return dp[n];
    }
};

int main(){
    Solution solution;

    int n = 25;
    cout << solution.tribonacci(n) << endl;
    return 0;
}

File: python/LeetCode Quiz/No1137.py
'''
The Tribonacci sequence Tn is defined as follows: 

T0 = 0, T1 = 1, T2 = 1, and Tn+3 = Tn + Tn+1 + Tn+2 for n >= 0.

Given n, return the value of Tn.
'''

class Solution:
    def tribonacci(self, n: int) -> int:
        mem = dict()
        def cal(n: int) -> int:
            if n<2:
                return n
            if n==2:
                return 1
            if n not in mem:
                mem[n] = cal(n-3) + cal(n-2) + cal(n-1)
            return mem[n]
        return cal(n)

solution = Solution()
n = 25
print(solution.tribonacci(n))